Support large files try 2

aliabid94 opened the pull request:
Rewrote large file issue fixer with @freddyaboulton's approach. When a file is uploaded to FileUpload, it is automatically sent to the backend, and then the frontend receives a reference to the name of this new file. Much cleaner, thanks @freddyaboulton!

Fixes: #2421

aliabid94 commented:
ready for review btw

abidlabs commented on gradio/routes.py:
```diff
@@ -509,6 +511,27 @@ async def join_queue(
+                else:
+                    prefix = input_file.filename
+                    suffix = ""
+                output_file_obj = tempfile.NamedTemporaryFile(delete=False, suffix=f".{suffix}", prefix=f"{prefix}_")
```
Nit: if `suffix` is `""`, this will add a `.` to the end of the temporary filename

Note, there might also be issues with the filename containing invalid characters, etc. I would reuse the implementation of `TempFileManager.get_prefix_and_extension()` (maybe make that a classmethod) where we have already handled this logic

abidlabs commented on test/test_routes.py:
```diff
@@ -54,6 +54,15 @@ def test_get_config_route(self, test_client):
+        assert response.status_code == 200
+        files = response.json()
+        with open(files[0]) as saved_file:
+            assert saved_file.read() == "abcdefghijklmnopqrstuvwxyz"
```
You may want to add a test to ensure that the temporary file name is what you expect

I also I would add a test containing a larger file. I guess it doesn't have to be massive but maybe more than 1024 bytes so that we know that the chunking is accurate.

abidlabs commented:
Really cool PR @aliabid94! I tested it locally and on Spaces 

**Locally, I ran the following demo:**

```py
import gradio as gr

def test(x):
    return x.name

with gr.Blocks() as demo:
    inp = gr.File(label="Input")
    vis_out=gr.File(label="output")
    with gr.Row():
        btn = gr.Button("Run")
    btn.click(fn=test, inputs=[inp], outputs=[vis_out])

demo.launch() 
```

**Feedback:**

* The UI looks good:
<img width="507" alt="image" src="https://user-images.githubusercontent.com/1778297/219270648-a54d22e9-bd17-42fb-98b8-472bacbbd155.png">
* The demo works well for small file sizes, except that the download link on the input file component is broken:

<img width="520" alt="image" src="https://user-images.githubusercontent.com/1778297/219272268-ddef0eda-5e71-4163-a1be-9c1e2f4a17d6.png">

Also I see this error in the js console but it doesn't seem to have a detrimental effect:

<img width="562" alt="image" src="https://user-images.githubusercontent.com/1778297/219273688-d3b79800-cac5-4a4d-97b9-b596f56663e1.png">

Details of the error:
```
{"detail":[{"loc":["body","files",0],"msg":"Expected UploadFile, received: <class 'str'>","type":"value_error"}]}
```

* However, if one clicks on the submit button while the `File` component is still uploading the file (which for me takes about 10 seconds for a 5 MB file by the way), then an empty file is sent to the backend, resulting in this:

<img width="520" alt="image" src="https://user-images.githubusercontent.com/1778297/219271142-2e8c7842-c3c1-4d5e-8ba5-ad785b73dc5a.png">

* As we discussed briefly in our sync, I think the `File` component would need to "block" the submit button from happening until the file has been fully uploaded. When someone tries to submit, I suppose we would need to "signal" that the file component is still uploading and that's why the event cannot start. This could be similar to whatever UI affordance we use to designate that a field is required (#2718) 
* As a side note, I don't find the orange border indicator as too visually obvious that the file is uploading. Perhaps the "Download" link could be replaced with "Uploading..." in italics until the file finishes uploading?
* For larger file sizes, it just takes *forever* to upload the files locally. I'd say it took about 5-10 minutes for a 200 MB file, which seems kind of weird since everything is local

**on Spaces, using the deployed PR link, I looked at the `zip_files` demo**:

* Generally the same feedback as above, although in the case of the `zip_files` demo, neither the input file nor the output file download link worked for me

**Other notes**

* Finally, I wonder if we should add a `max_size` parameter to govern the maximum file size that someone can submit and set it to a reasonably small default like 100 MB. Otherwise, given the relatively small disk space on Spaces (50GB), it could easily get filled up

----

The critical questions imo are:
* why is the upload speed so slow?
* why are the download links broken?
* how do we prevent users from triggering events before the file has finished uploading?
 
Happy to give this another look!

freddyaboulton commented:
> I think the File component would need to "block" the submit button from happening until the file has been fully uploaded.

Without this PR, users can still hit submit with missing components and get errors, so I don't think we should increase the scope of this PR to include #2718.

I agree that we should make the "uploading" state more obvious though! Perhaps we can grey out the component and add "Uploading ..." with a spinner?

This is a really awesome PR @aliabid94 ! I agree with all of @abidlabs other comments. The one thing I noticed is that with cached examples, the file names change when you hit `Submit` 

![upload_cache_file_change_name](https://user-images.githubusercontent.com/41651716/219421460-3f9a02ee-8167-4c50-86fa-1d533970cc05.gif)

abidlabs commented:
> Without this PR, users can still hit submit with missing components and get errors, so I don't think we should increase the scope of this PR to include https://github.com/gradio-app/gradio/issues/2718.

Makes sense, agreed. Then it's just important to make sure the component makes it very clear visually when it's in the uploading state. 

> The one thing I noticed is that with cached examples, the file names change when you hit Submit

This is actually an older issue I believe, not introduced in this PR. I see it here as well: https://huggingface.co/spaces/gradio-pr-deploys/pr-3149-all-demos

aliabid94 commented:
Ok made the following changes:
- should upload much faster! (~200MB/s locally for me) problem was I was reading the stream 1 kb at a time, now reading 100 MB at a time. 
- fixed console error issue
- uploaded files are downloadable
  - added an uploaded_file dir that keeps uploaded files
- use TempFileManager in uploads
- Shows "Uploading..." while uploading. Since it's so fast now, not too urgent to make it look better.

will set max size and block submits on upload in another PR.

abidlabs commented on gradio/routes.py:
```diff
@@ -292,17 +295,16 @@ async def file(path_or_url: str, request: fastapi.Request):
                     for dir in blocks.file_directories
                 )
             )
+            was_uploaded = abs_path.startswith(app.uploaded_file_dir)
```
Is `startswith()` robust? I think using the `in` operator with two `Path`s is more robust

aliabid94 replied:
No, it's the opposite - the `in` keyword is not robust. For example, if the upload dir is "a/b", and a user tries to access file "c/d/a/b", the `in` keyword will allow access to this. This is not secure. We should modify the other lines to use `startswith` as well - I can do this is in this PR

abidlabs replied:
Not `in` with two strings, but `in` with two `pathlib.Path` objects. See here: https://stackoverflow.com/a/34236245/5209347

aliabid94 replied:
oh yep you're right

abidlabs commented on gradio/routes.py:
```diff
@@ -292,17 +295,16 @@ async def file(path_or_url: str, request: fastapi.Request):
-            created_by_app = str(utils.abspath(path_or_url)) in set().union(
-                *blocks.temp_file_sets
-            )
+            created_by_app = abs_path in set().union(*blocks.temp_file_sets)
```
Thanks for fixing

abidlabs commented:
Beautiful, LGTM @aliabid94! I left a couple of nits on the backend code and I'd suggest @pngwn or @dawoodkhan82 do another pass on the svelte, but works beautifully for me!

pngwn reviewed (approved):
Frontend changes look good to me.
